internal/executor/executor.go

```diff
@@ -201,9 +202,38 @@ func (e *Executor) runSingleTask(ctx context.Context, task *build.Task) (err err
 		}
 	}()
 
+	// Monitor heartbeats and cancel the task if we don't receive
+	// them for more than maxNoHeartbeatsDuration
+	const maxNoHeartbeatsDuration = 2 * time.Minute
+
+	currentTime := time.Now()
+	task.LastHeartbeatReceivedAt.Store(&currentTime)
+
+	go func() {
+		for {
+			if time.Since(*task.LastHeartbeatReceivedAt.Load()) > maxNoHeartbeatsDuration {
+				cancel()
+			}
+
+			select {
+			case <-time.After(time.Second):
+				continue
+			case <-ctx.Done():
+				break
+			}
+		}
+	}()
+
+	// Run the task
 	if err := task.Instance.Run(ctx, &instanceRunOpts); err != nil {
 		switch {
+		case errors.Is(ctx.Err(), context.Canceled):
```

Comment:
Would it also be triggered if the parent context is canceled?

Reply:
Indeed, fixed this in https://github.com/cirruslabs/cirrus-cli/pull/652/commits/bba2f3f03839257e953cadd0a6a8b345f826c0fc.